src/renderers/shared/reconciler/ReactCompositeComponent.js

```diff
@@ -368,18 +369,22 @@ var ReactCompositeComponentMixin = {
    * @final
    * @internal
    */
-  unmountComponent: function() {
+  unmountComponent: function(safely) {
     if (!this._renderedComponent) {
       return;
     }
     var inst = this._instance;
 
     if (inst.componentWillUnmount) {
-      inst.componentWillUnmount();
+      if (safely) {
+        ReactErrorUtils.invokeGuardedCallback('componentWillUnmount', inst.componentWillUnmount.bind(inst));
```

Comment:
This codepath is a production codepath, so I was thinking we should avoid doing string concatenation unnecessarily.  If we want to avoid the concatenation, we can only have "componentWillUnmount" or the component name, but not both, right?  I don't feel strongly, I'd be happy to do any of the three (concat, functionName, className), just let me know which.  We could also fix in a followup if we decide it would be necessary/helpful.